From dc8be8046fa181f5babb39ce6dac89a8dc6432bd Mon Sep 17 00:00:00 2001
From: Afzal Mohammed <afzal@ti.com>
Date: Mon, 27 Feb 2012 19:22:54 +0530
Subject: [PATCH 0820/1092] ARM: OMAP2+: gpmc: nand support

gpmc driver probe will check the capability of the
devices passed via platform data. Upon finding nand
device, gpmc nand initialization would be done. The
initialization function will internally create
platform device for nand.

Signed-off-by: Afzal Mohammed <afzal@ti.com>
---
 arch/arm/mach-omap2/gpmc-nand.c | 2 +-
 arch/arm/mach-omap2/gpmc.c      | 6 ++++++
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-omap2/gpmc-nand.c b/arch/arm/mach-omap2/gpmc-nand.c
index 2de65d1..c12b286 100644
--- a/arch/arm/mach-omap2/gpmc-nand.c
+++ b/arch/arm/mach-omap2/gpmc-nand.c
@@ -81,7 +81,7 @@ static int omap2_nand_gpmc_retime(struct omap_nand_platform_data *gpmc_nand_data
 	return 0;
 }
 
-int __init gpmc_nand_init(struct omap_nand_platform_data *gpmc_nand_data)
+int __devinit gpmc_nand_init(struct omap_nand_platform_data *gpmc_nand_data)
 {
 	int err	= 0;
 	u8 cs = 0;
diff --git a/arch/arm/mach-omap2/gpmc.c b/arch/arm/mach-omap2/gpmc.c
index c11c4fa..ff38a55 100644
--- a/arch/arm/mach-omap2/gpmc.c
+++ b/arch/arm/mach-omap2/gpmc.c
@@ -29,6 +29,7 @@
 
 #include <asm/mach-types.h>
 #include <plat/gpmc.h>
+#include <plat/nand.h>
 
 #include <plat/sdrc.h>
 
@@ -747,6 +748,8 @@ static int __devinit gpmc_probe(struct platform_device *pdev)
 	u32 l;
 	int ret = -EINVAL;
 	struct resource *res = NULL;
+	struct gpmc_devices_info *gpmc_device = pdev->dev.platform_data;
+	void *p;
 
 	/* XXX: This should go away with HWMOD & runtime PM adaptation */
 	gpmc_clk_init();
@@ -789,6 +792,9 @@ static int __devinit gpmc_probe(struct platform_device *pdev)
 
 	gpmc_mem_init();
 
+	for (p = gpmc_device->pdata; p; gpmc_device++, p = gpmc_device->pdata)
+		if (gpmc_device->flag & GPMC_DEVICE_NAND)
+			gpmc_nand_init((struct omap_nand_platform_data *) p);
 	return 0;
 
 err_remap:
-- 
1.7.11.2

